src/Matrix.jl

```diff
@@ -18,12 +18,6 @@ base_ring_type(::Type{<:MatrixElem{T}}) where T <: NCRingElement = parent_type(T
 
 base_ring(a::MatrixElem{T}) where {T <: NCRingElement} = a.base_ring::parent_type(T)
 
-function check_parent(a::MatElem, b::MatElem, throw::Bool = true)
```

Comment:
Is this one of the situations where we want to overwrite it?

Reply:
Input:
```julia
function new_check_parent(a, b, throw::Bool = true)
   flag = parent(a) === parent(b)
   flag || !throw || error("parents do not match")
   return flag
end

function old_check_parent(a::MatElem, b::MatElem, throw::Bool = true)
  fl = (base_ring(a) != base_ring(b) || nrows(a) != nrows(b) || ncols(a) != ncols(b))
  fl && throw && error("Incompatible matrix spaces in matrix operation")
  return !fl
end

R,(x,y) = QQ[:x, :y]
M = [
  matrix(ZZ, [1 2 ; 3 4]),
  matrix(R, [1 2 ; 3 4]),
  matrix(R, [0 1 2 ; 3 4 5]),
  matrix(QQ, [1//2 2//3 ; 3//4 4//5]),
  matrix(QQ, [3//2 2//3 ; 3//4 4//5]),
  matrix(QQ, [3//2 2//3 1 ; 3//4 4//5 0]),
]

for i in 1:length(M), j in i:length(M)
  println(i,j)
  @btime new_check_parent($(M[i]), $(M[j]), false)
  @btime old_check_parent($(M[i]), $(M[j]), false)
end
```

Timings are identical across the board.

Reply:
For me, they are not identical across the board:
```
julia> @btime new_check_parent($(M[1]), $(M[4]), false)
  3.464 ns (0 allocations: 0 bytes)
false

julia> @btime old_check_parent($(M[1]), $(M[4]), false)
  1.584 ns (0 allocations: 0 bytes)
false
```

Reply:
OK, it seems there is a difference between QQ from AbstractAlgebra and QQ from Nemo... I was purely measuring in `AbstractAlgebra` and I just re-did that, and I get:
```
julia> @btime new_check_parent($(M[1]), $(M[4]), false)
  1.833 ns (0 allocations: 0 bytes)
false

julia> @btime old_check_parent($(M[1]), $(M[4]), false)
  1.833 ns (0 allocations: 0 bytes)
false
```

But if I instead start with `using Nemo` and execute the test there, I get:
```
julia> @btime new_check_parent($(M[1]), $(M[4]), false)
  1.500 ns (0 allocations: 0 bytes)
false

julia> @btime old_check_parent($(M[1]), $(M[4]), false)
  0.875 ns (0 allocations: 0 bytes)
false
```

Inspecting the code Julia generates, for `old_check_parent` it basically compiles the function down to `return false` because the base rings are both singletons and different. (Though now I don't quite understand why this doesn't also kick in for e.g. 1+2 or 3+4 or other such combos.. huh

Anyway, thanks for checking carefully. I'll restore this method.

Reply:
Thanks

Reply:
BTW I figured out that if one changes it to first check the parent *type*, i.e.
```
function new_check_parent(a, b, throw::Bool = true)
   flag = parent_type(a) === parent_type(b) && parent(a) === parent(b)
   flag || !throw || error("parents do not match")
   return flag
end
```
it gets faster in all those cases I was wondering about, i.e., where the parents obviously don't match. Adding this in type *stable* code incurs no extra cost as the new check is always optimized to either `true` or `false`. But in type unstable code it is one extra step. Also, the only benefit is for code where the parents "obviously" have mismatching parents. I am not sure this is worth optimizing, as I think there is far less code that legitimately calls `check_parent` on objects where the parents don't even have matching type, compared to code which is for some direct or indirect (!) reason evaluated in a type unstable fashion. (Also, of course in the end we are talking about *nanoseconds* here).

Anyway, at least now my curiosity is satisfied: apparently the Julia compiler is just not quite bright enough to (always?) prove that `a===b` never holds when `typeof(a)!==typeof(b)`.